### PR TITLE
Day 3 registration with db connection

### PR DESCRIPTION
**Pull request by satya-rayi**
Day 3 registration with db connection

### CODE REVIEW

File: application.py [hunk 1]
```diff
@@ -27,12 +28,21 @@
...
+    users = db.query(User).all()
+    return render_template('users.html',users=users)  
+
 @app.route('/register', methods=['GET', 'POST']) 
 def register():
     if request.method == 'POST':
-         username = request.form['username']
-         password = request.form["password"]
-         return render_template('welcome.html',username=username)
+        import time
```
Comment (praveengarimella) [inline application.py]: Move it to the top of the file

File: application.py [hunk 2]
```diff
@@ -27,12 +28,21 @@
...
-         username = request.form['username']
-         password = request.form["password"]
-         return render_template('welcome.html',username=username)
+        import time   
+        username = request.form['username']
+        password = request.form["password"]
+        user = User(username=username, password=password, created_at=time.strftime('%Y-%m-%d %H:%M:%S'))
+        db.add(user)
+        db.commit()
+        return render_template('welcome.html',username=username)
```
Comment (praveengarimella) [inline application.py]: What happens when it is a duplicate username?

File: application.py [hunk 3]
```diff
@@ -27,12 +28,21 @@
...
 @app.route('/register', methods=['GET', 'POST']) 
 def register():
     if request.method == 'POST':
-         username = request.form['username']
-         password = request.form["password"]
-         return render_template('welcome.html',username=username)
+        import time   
+        username = request.form['username']
+        password = request.form["password"]
+        user = User(username=username, password=password, created_at=time.strftime('%Y-%m-%d %H:%M:%S'))
```
Comment (praveengarimella) [inline application.py]: How about setting the time as a default value of the database instead of setting it here.

File: database_setup.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+#create declarative_base instance
+Base = declarative_base()
+
+#we'll add classes here
+class User(Base):  
+    __tablename__ = 'user'
+    id = Column(Integer, Sequence('id', start=1, increment=1),primary_key=True)    
+    username = Column(String)
+    password = Column(String)
+    created_at = Column(DateTime)
```
Comment (praveengarimella) [inline database_setup.py:29]: Can this be taken as a default value, i.e., the current timestamp in the database?

File: application.py (hunk 2 above)
Comment (satya-rayi) [reply application.py]: writing that code...,

File: templates/welcome.html [hunk 5]
```diff
@@ -18,7 +18,7 @@
       
           <!-- Login Form -->
           <div style="margin:5%;">
-            Welcome : {{username}}
+            Welcome : {{username}}, You are successfully Registered
           </div>
           <div><a href="register">Register</a></div>
```
Comment (praveengarimella) [inline templates/welcome.html]: Can we avoid the code duplication by using the templates better? For example, all the HTML and Head tags can be part of a generic template which can be imported?

**Review by praveengarimella — CHANGES_REQUESTED** [review]
Pls see my comments, fix them and then merge them into the master branch.

File: application.py (hunk 1 above)
Comment (satya-rayi) [reply application.py]: done

File: application.py (hunk 2 above)
Comment (satya-rayi) [reply application.py]: if it is duplicate user name, a message will be displayed saying so.

File: templates/welcome.html (hunk 5 above)
Comment (satya-rayi) [reply templates/welcome.html]: sure, will do that..

File: application.py [hunk 6]
```diff
@@ -30,18 +33,23 @@ def index():
...
-        import time   
+    if request.method == 'POST':        
         username = request.form['username']
         password = request.form["password"]
-        user = User(username=username, password=password, created_at=time.strftime('%Y-%m-%d %H:%M:%S'))
-        db.add(user)
-        db.commit()
+        password = hashlib.md5(password.encode()).hexdigest()
+        
+        exists = db.query(User.id).filter_by(username=username).scalar()
```
Comment (praveengarimella) [inline application.py:46]: Can it be done by catching an exception when the insert fails? That way you will reduce one request to the database.

**Review by praveengarimella — COMMENTED** [review]
Looks good, but I've left one more comment for you to explore.

File: application.py (hunk 6 above)
Comment (satya-rayi) [reply application.py:46]: It can be, will do that

File: application.py (hunk 3 above)
Comment (satya-rayi) [reply application.py]: okay

File: database_setup.py (hunk 4 above)
Comment (satya-rayi) [reply database_setup.py:29]: yes

File: templates/welcome.html (hunk 5 above)
Comment (satya-rayi) [reply templates/welcome.html]: done this one..

**Pull request merged**